Author: aszlig <aszlig@nix.build>
Date:   Tue Mar 27 05:37:13 2018 +0200

    Only allow users with a valid invitation link
    
    We don't want everybody to create accounts on this instance but only
    selected users we explicitly invite, so let's remove the registration
    form from the home page and only show it whenever the user directly
    follows an invitation URL.
    
    Signed-off-by: aszlig <aszlig@nix.build>
    Filename: invite-only.patch

diff --git a/website/client/components/static/home.vue b/website/client/components/static/home.vue
index 018018077..265ff3d70 100644
--- a/website/client/components/static/home.vue
+++ b/website/client/components/static/home.vue
@@ -5,7 +5,7 @@
       a(href='http://www.enable-javascript.com/', target='_blank') {{ $t('jsDisabledLink') }}
 
     #intro-signup.purple-1
-      .container
+      .container(v-if='invited')
         .row
           .col-12.col-md-6.col-lg-6
             img(src='~assets/images/home/home-main@3x.png', width='357px')
@@ -24,6 +24,12 @@
               button.sign-up(@click="register()") {{$t('signup')}}
           .col-12
             .spacer.svg-icon(v-html='icons.spacer')
+      .container(v-if='!invited')
+        img(src='~assets/images/home/home-main@3x.png', width='357px')
+        h1 {{$t('motivateYourself')}}
+        p.section-main {{$t('timeToGetThingsDone', {userCountInMillions})}}
+        .col-12
+          .spacer.svg-icon(v-html='icons.spacer')
 
     #gamify-life.purple-2
       .container-fluid
@@ -543,6 +549,17 @@
         if (this.passwordConfirm.length <= 3) return false;
         return this.passwordConfirm !== this.password;
       },
+      invited () {
+        let groupInvite = '';
+        if (this.$route.query && this.$route.query.p) {
+          groupInvite = this.$route.query.p;
+        }
+
+        if (this.$route.query && this.$route.query.groupInvite) {
+          groupInvite = this.$route.query.groupInvite;
+        }
+        return Boolean(groupInvite);
+      },
     },
     watch: {
       username () {
diff --git a/website/server/libs/auth/index.js b/website/server/libs/auth/index.js
index 2a34ef05c..454146c40 100644
--- a/website/server/libs/auth/index.js
+++ b/website/server/libs/auth/index.js
@@ -50,7 +50,9 @@ async function _handleGroupInvitation (user, invite) {
     }
   } catch (err) {
     logger.error(err);
+    return false;
   }
+  return true;
 }
 
 function hasLocalAuth (user) {
@@ -141,7 +143,11 @@ async function registerLocal (req, res, { isV3 = false }) {
 
   // we check for partyInvite for backward compatibility
   if (req.query.groupInvite || req.query.partyInvite) {
-    await _handleGroupInvitation(newUser, req.query.groupInvite || req.query.partyInvite);
+    let success = await _handleGroupInvitation(newUser, req.query.groupInvite || req.query.partyInvite);
+    if (!success)
+      throw new NotAuthorized(res.t('inviteOnly'));
+  } else {
+    throw new NotAuthorized(res.t('inviteOnly'));
   }
 
   let savedUser = await newUser.save();
